fl4health/model_bases/autoencoders_base.py

```diff
@@ -0,0 +1,136 @@
+    All we need to define such model is the type of the model, and the structure
+    of the encoder and the decoder modules. This type of model should have the capability
+    to encode data using the encoder module and decode the output of the encoder using the decoder module.
+    """
+
+    def __init__(
+        self,
+        encoder: nn.Module,
+        decoder: nn.Module,
+    ) -> None:
+        super().__init__()
+        self.encoder = encoder
+        self.decoder = decoder
+
+    @abstractmethod
+    def forward(self, input: torch.Tensor) -> torch.Tensor:
+        # Forward is called in client classes with a single input tensor.
+        raise NotImplementedError
+
+
+class BasicAe(AbstractAe):
+    def __init__(self, encoder: nn.Module, decoder: nn.Module) -> None:
+        super().__init__(encoder, decoder)
+
+    def encode(self, input: torch.Tensor) -> torch.Tensor:
+        latent_vector = self.encoder(input)
+        return latent_vector
+
+    def decode(self, latent_vector: torch.Tensor) -> torch.Tensor:
+        output = self.decoder(latent_vector)
+        return output
+
+    def forward(self, input: torch.Tensor) -> torch.Tensor:
+        z = self.encode(input)
+        return self.decode(z)
+
+
+class VariationalAe(AbstractAe):
+    """Variational Auto-Encoder model base class."""
+
+    def __init__(
+        self,
+        encoder: nn.Module,
+        decoder: nn.Module,
+    ) -> None:
+        """Initializes the VariationalAE class.
+
+        Args:
+            model_type (AutoEncoderType): The model type should be VARIATIONAL_AE.
+            encoder (nn.Module): Encoder module defined by the user.
+            decoder (nn.Module): Decoder module defined by the user.
+        """
+        super().__init__(encoder, decoder)
+
+    def encode(self, input: torch.Tensor) -> Tuple[torch.Tensor, torch.Tensor]:
+        mu, logvar = self.encoder(input)
+        return mu, logvar
+
+    def decode(self, latent_vector: torch.Tensor) -> torch.Tensor:
+        output = self.decoder(latent_vector)
+        return output
+
+    def sampling(self, mu: torch.Tensor, logvar: torch.Tensor) -> torch.Tensor:
+        std = torch.exp(0.5 * logvar)
+        eps = torch.randn_like(std)
+        return eps.mul(std).add_(mu)
+
+    def forward(self, input: torch.Tensor) -> torch.Tensor:
+        mu, logvar = self.encode(input)
+        z = self.sampling(mu, logvar)
+        output = self.decode(z)
+        # Output (reconstruction) is flattened to be concatenated with mu and logvar vectors.
+        # The shape of the flattened_output can be later restored by having the training data shape,
+        # or the decoder structure.
+        flattened_output = output.view(output.shape[0], -1)
+        return torch.cat((logvar, mu, flattened_output), dim=1)
+
+
+class ConditionalVae(AbstractAe):
+    def __init__(
+        self, encoder: nn.Module, decoder: nn.Module, converter: Optional[AutoEncoderDatasetConverter] = None
```

Comment:
We should be able to construct a callable with the data_shape and get_condition_vector_size() using the object. That callable could be started instead of the object. We can talk about this a bit if you want. The idea would be to use pythons `partial` functionality.

Reply:
Ah right, that's a great idea!

Reply:
It is done by adding the `get_unpacking_function()` in the `dataset_converter` class.

Reply:
TL;DR: I ended up not using `partial` and fixed it in another way. Yet, it is just sending a function rather than the whole object.

Details:
I had used `partial` and all the examples were working just fine, but to make it work, I had changed the order of clients' setup (in `BasicClient`) to first initiate dataloaders and then initiate the model (to be able to use info from the data (data shape) in the model construction). But then results would be different for a given random seed. For smoke tests, the random seed used to check the results is 42, and with this change in order of operations, the previous results for that random seed would be slightly different now (and smoke tests failed). So I had to find a way in which I don't need to change the order of operations in client setup.